src/aics-image-viewer/components/Toolbar/index.tsx

```diff
@@ -41,116 +42,214 @@ interface ToolbarProps {
   };
 }
 
-export default function Toolbar(props: ToolbarProps): React.ReactElement {
-  const { renderConfig, showAxes, showBoundingBox, autorotate } = props;
+interface ToolbarState {
+  scrollMode: boolean;
+  scrollBtnLeft: boolean;
+  scrollBtnRight: boolean;
+}
+
+export default class Toolbar extends React.Component<ToolbarProps, ToolbarState> {
+  barRef: React.RefObject<HTMLDivElement>;
+  leftRef: React.RefObject<HTMLDivElement>;
+  rightRef: React.RefObject<HTMLDivElement>;
+  centerRef: React.RefObject<HTMLDivElement>;
+
+  constructor(props: ToolbarProps) {
+    super(props);
+    this.state = {
+      scrollMode: false,
+      scrollBtnLeft: false,
+      scrollBtnRight: false,
+    };
+    this.barRef = React.createRef();
+    this.leftRef = React.createRef();
+    this.rightRef = React.createRef();
+    this.centerRef = React.createRef();
+
+    window.addEventListener("resize", this.checkSize);
+    this.checkSize();
+  }
 
-  const twoDMode = props.mode !== ViewMode.threeD;
+  checkSize = debounce((): void => {
+    const { leftRef, centerRef, rightRef, barRef } = this;
+    const leftRect = leftRef.current!.getBoundingClientRect();
+    const centerRect = centerRef.current!.getBoundingClientRect();
+    const rightRect = rightRef.current!.getBoundingClientRect();
 
-  const renderGroup1 =
-    renderConfig.viewModeRadioButtons || renderConfig.resetCameraButton || renderConfig.autoRotateButton;
-  const renderGroup4 = renderConfig.showAxesButton || renderConfig.showBoundingBoxButton;
+    if (this.state.scrollMode) {
+      // Leave scroll mode if there is enough space for centered controls not to overlap left/right-aligned ones
+      const barWidth = barRef.current!.getBoundingClientRect().width;
+      const requiredWidth = Math.max(leftRect.width, rightRect.width) * 2 + centerRect.width + 30;
+      if (barWidth > requiredWidth) {
+        this.setState({ scrollMode: false });
+      }
+    } else {
+      // Enter scroll mode if centered controls are overlapping either left/right-aligned ones
+      if (leftRect.right > centerRect.left || centerRect.right > rightRect.left) {
+        this.setState({ scrollMode: true });
+      }
+    }
+    this.checkScrollBtnVisible();
+  }, 50);
```

Comment:
This 50 could also be a named constant.  Although its meaning is a bit more self evident, it's not clear why it should be 50.

Reply:
Just to be sure I understand, is this rate-limiting the checkSize calls to no more than 1 every 50ms ?

Reply:
Yeah, `_.debounce` delays calling a function until it hasn't been called in the last 50 (or whatever) ms. This is actually the second time I've used `debounce` with a limit of 50 - it's a totally arbitrary choice, but could be worth its own entry in constants.ts?

Reply:
yes or right here in this file, jsut give it a name like CHECKSIZE_DEBOUNCE_RATE or whatever